### PR TITLE
Implemented I18N::normalizeLocale() that noramlizes ICU or BCP47 locale into Yii translation language

### PR DESCRIPTION
**Pull request by samdark**
Part of #7883 fix.

TODO:
- [x] Better method name?
- [x] Introduce method for getting fallback language to take script into account.
- [x] Adjust `BaseFileHelper::localize()`.
- [x] Adjust JUI DatePicker asset (https://github.com/yiisoft/yii2-jui/pull/15)
- [ ] Adjust `DbMessageSource`.
- [ ] Adjust `PhpMessageSource`.
- [ ] Adjust `GettextMessageSource`.

### CODE REVIEW

File: framework/i18n/I18N.php [hunk 1]
```diff
@@ -199,4 +199,86 @@ public function getMessageSource($category)
...
+     * - Language tag according to https://www.rfc-editor.org/bcp/bcp47.txt
+     * - ICU locale string http://userguide.icu-project.org/locale
+     *
+     * @return string normalized language ID
+     *
+     * @since 2.0.4
+     */
+    public function normalizeLocale($locale)
+    {
+        if (!class_exists('\Locale')) {
```
Comment (cebe) [inline framework/i18n/I18N.php]: should not trigger autoload if it does not exist.

**Comment by cebe** [comment]
Looks good to me and also fine with the name.

File: framework/i18n/I18N.php [hunk 2]
```diff
@@ -199,4 +199,108 @@ public function getMessageSource($category)
...
+     * @param string $language language string. Could be either:
+     *
+     * - Language tag according to https://www.rfc-editor.org/bcp/bcp47.txt
+     * - ICU locale string http://userguide.icu-project.org/locale
+     *
+     * @return string normalized language ID
+     *
+     * @since 2.0.4
+     */
+    protected function fallbackNormalizeLocale($locale)
```
Comment (klimov-paul) [inline framework/i18n/I18N.php]: Are you sure it should be 'protected'? Why not 'private'?

Comment (cebe) [inline framework/i18n/I18N.php]: I think it should be private too, this is only a fallback if intl is not available which should not be exposed.

File: framework/i18n/I18N.php [hunk 3]
```diff
@@ -199,4 +199,108 @@ public function getMessageSource($category)
...
+     * "zh-Hant-CN" ? "zh-CN" ? "zh"
+     *
+     * @param string $languageId normalized language ID. Either "zh-Hant-CN", "zh-CN" or "CN".
+     * @return string fallback language ID or null if there's no fallback
+     */
+    public function getFallbackLanguageId($languageId)
+    {
+        $parts = explode('-', $languageId);
+
+        switch(count($parts)) {
```
Comment (klimov-paul) [inline framework/i18n/I18N.php]: Missing space after 'switch'

File: framework/i18n/I18N.php [hunk 4]
```diff
@@ -199,4 +199,108 @@ public function getMessageSource($category)
 
         throw new InvalidConfigException("Unable to locate message source for category '$category'.");
     }
+
+    /**
+     * Normalizes language string to "zh-Hant-CN", "zh-CN" or "CN" format ready
+     * to be used to specify translation language.
+     *
+     * @param string $language language string. Could be either:
```
Comment (klimov-paul) [inline framework/i18n/I18N.php]: Param name missmatch: '$language' at docs vs '$locale' at definition.

File: tests/framework/i18n/ExposedI18N.php [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * @package yiiunit\framework\i18n
+ */
+class ExposedI18N extends I18N
+{
+    /**
+     * @inheritdoc
+     */
+    public function fallbackNormalizeLocale($locale)
+    {
+        return parent::fallbackNormalizeLocale($locale);
```
Comment (klimov-paul) [inline tests/framework/i18n/ExposedI18N.php]: Bypass via Reflection can be used instead.

File: framework/i18n/I18N.php [hunk 6]
```diff
@@ -199,4 +199,108 @@ public function getMessageSource($category)
...
+     *
+     * - Language tag according to https://www.rfc-editor.org/bcp/bcp47.txt
+     * - ICU locale string http://userguide.icu-project.org/locale
+     *
+     * @return string normalized language ID
```
Comment (klimov-paul) [inline framework/i18n/I18N.php]: Perhaps doc comment should use 'locale' term instead of 'language'?

**Comment by klimov-paul** [comment]
> Adjust BaseFileHelper::localize().

Whould not this break a BC?
Developer may not stumble on the #7883 issue, while he might plac his localized files under 'en_us' instead of 'en-us'. Change will make them unavailbale.

File: tests/framework/i18n/ExposedI18N.php (hunk 5 above)
Comment (samdark) [inline tests/framework/i18n/ExposedI18N.php]: How?

**Comment by samdark** [comment]
Yes, that could break things in that case.

File: tests/framework/i18n/ExposedI18N.php (hunk 5 above)
Comment (klimov-paul) [inline tests/framework/i18n/ExposedI18N.php]: See for example:
https://github.com/yiisoft/yii2-bootstrap/pull/31/files#diff-2f1760bba264cf4f4b0167dbbe1599e3R21

**Comment by samdark** [comment]
Won't be able to finish it before 2.0.4. Moving to 2.0.5.

**Comment by samdark** [comment]
Adjusted PR according to suggestions but not sure how to deal with BC in this case...

**Comment by samdark** [comment]
I'm for adding it into upgrade notes and proceeding with merging it. Any objections?

**Comment by cebe** [comment]
not sure, maybe it is time to start the 2.1 branch?

**Comment by PowerGamer1** [comment]
@samdark 
https://github.com/yiisoft/yii2-jui/pull/15/files#diff-b21f181a007dbc219c5eca1f1e93d2d0R46

You normalize the language specified by user for the widget (`$this->language`) but not the language in the filename - where is the logic in this? In particular, your code won't find "/ui/i18n/datepicker-en_US.js" file if the user specifies "en_US" language for the widget.

**Comment by samdark** [comment]
Won't be merged into 2.0.

**Pull request closed without merging**